terminal: Implement own context handler for copy&paste

marusak opened the pull request:
Implementing own context menu when right click on terminal. Support copy
and paste even when screenReaderMode (accessibility mode) is used.

To see demo watch https://youtu.be/OHdrFrMSgEg

Context menu changed a bit from the demo, current screenshot:
![currectscreen](https://user-images.githubusercontent.com/12330670/52548648-2c867400-2dcf-11e9-8d3a-367937d8b361.png)



## Browsers support
### Chrome
Full support since version 66. On first paste will ask for permission.

### Firefox
Copy support since version 63. Paste does not work :( Firefox did not implement clipboard permissions for web applications (only for extensions).

### Opera
Full support since version 53. On first paste will ask for permission.

### Edge
Not tested, but seems there is no support at all.

## Using sync vs async
This PR implements newer async clipboard API. There is also sync approach by using method ` execCommand` which is deprecated but is supported from much older version in all browsers. However pasting in firefox is still not supported even with this older method.

## Pros

- Copy&paste
- Discoverable keyboard shortcuts

## Cons

- Paste does not work in all browsers
- Edge is (most likely) not supported at all
- Works only in newer versions of browsers (possibility to use old deprecated approach?) 

Note: There are no tests and that't for two reasons:

1. I want to be sure this will get accepted before I fight with tests ( as there are some issues with some browsers)
2. I am not sure where and how to write tests for this, will need some guidance.

TODO:
 - [x] fix text selection in test
 - [x] cover keyboard shortcuts (Ctrl+Insert and Shift+Insert) in test

martinpitt commented on pkg/lib/cockpit-components-context-menu.jsx:
```diff
@@ -0,0 +1,98 @@
+        }
+
+        if (bottom) {
+            this.root.style.top = `${clickY - rootH - 5}px`;
```
This looks like CSS attributes? Doesn't that trigger a Content Security Policy error?

marusak replied:
I have not noticed that CSP would happen. Most likely because it is only set when component is created.

martinpitt replied:
It's less a question of when it happens, but where -- Neither HTML or JS is usually allowed to inject CSS. But if it works, fine (I don't claim to understand all details of CSP).

martinpitt commented on pkg/lib/cockpit-components-context-menu.jsx:
```diff
@@ -0,0 +1,98 @@
+        return this.state.visible &&
+            <div ref={ ref => { this.root = ref } } className="contextMenu">
+                <div className="contextMenuOption" onClick={this.props.getText}>
+                    <div className="contextMenuName">Copy</div>
```
The menu entries need to be translated; import `cockpit` for that.

martinpitt commented on pkg/lib/cockpit-components-context-menu.jsx:
```diff
@@ -0,0 +1,98 @@
+                </div>
+            </div>;
+    }
+}
```
Please declare and document the properties of this component, especially because it is in pkg/lib.

martinpitt replied:
Declaring the properties still applies. Look at `git grep propTypes` for examples.

marusak replied:
aha, I misunderstood what you meant. Will fix that.

martinpitt commented on pkg/lib/cockpit-components-terminal.jsx:
```diff
@@ -130,6 +138,37 @@ export class Terminal extends React.Component {
+            navigator.clipboard.readText()
+                    .then((text) => {
+                        this.props.channel.send(text);
+                    })
```
Please let's avoid the unnecessary `{ ; }` if the arrow function body is a single expression (like in all three cases here).

martinpitt commented on pkg/lib/cockpit-components-terminal.jsx:
```diff
@@ -130,6 +138,37 @@ export class Terminal extends React.Component {
+                        this.props.channel.send(text);
+                    })
+                    .catch(() => {
+                        console.error('Text could not be pasted, use Shift+Insert');
```
Let's write the actual error as well. Or is it always clear (between multiple browsers) why this fails?

martinpitt commented on pkg/lib/cockpit-components-terminal.jsx:
```diff
@@ -130,6 +138,37 @@ export class Terminal extends React.Component {
+                        this.state.terminal.focus();
+                    });
+        } catch (error) {
+            console.error('Text could not be pasted, use Shift+Insert');
```
Why does this need to be caught twice?

martinpitt replied:
Answer to myself: I've seen this fail on Firefox with the "outer" exception, as the method call fails due  to the method not existing. But the returned promise fails in chromium if it doesn't have permissions, so it really can fail both ways.

marusak commented:
Thank you for review @martinpitt, I will update the source.
But I've been trying to do the test and some help would be really appreciated.
I have following code (at the end of test/verify/check-terminal)
```
          # Execute command                                                       
          b.key_press('echo "XYZ"\r')                                                 
          wait_line(n + 1, "XYZ")                                               
                                                                                                                               
          sel = line_sel(n + 1)                                                   
                                                                                  
          # Select line                                                           
          b.mouse(sel, "mousedown", 0, 0)                                         
          b.mouse(sel, "mouseup", 100, 0) # just random number for now                                        
                                                                                  
          # Right click and pick copy
          b.mouse(sel, "click", 0, 0, 2) # DOES NOT SHOW CONTEXT MENU             
          b.wait_present('.contextMenu')                                          
          b.wait_visible('.contextMenu')                                           
          b.click('.contextMenu .contextMenuOption:first-child')                  
                                                                                  
          # Right click and pick paste                                            
          b.mouse(sel, "click", 0, 0, 2) # DOES NOT SHOW CONTEXT MENU             
          b.wait_present('.contextMenu')                                          
          b.wait_visible('.contextMenu')                                          
          b.click('.contextMenu .contextMenuOption:nth-child(2)') # WILL ASK FOR PERMISSION TO PASTE
                                                                                  
          # Wait for text to show up                                              
          wait_line(n + 2, "XYZ")   
```
Firstly line is not selected (but I did not spend that much time debugging this part). But my main problem is that mouse click does not show the context menu. If I watch the test and do right click, test happily continues. And third problem is that the browser asks for permission to paste. Unless it is approved, it wont paste anything.
Can you give me some hints, please? Thank you!

martinpitt commented:
- I didn't yet look at the text selection. Presumably it needs some extra mousemove/mouseenter/mouseleave etc. events in addition to the down and up?

- Context menu works when using the [contextmenu](https://developer.mozilla.org/en-US/docs/Web/Events/contextmenu) mouse event instead of `click`, as that's what your code listens to:

      b.mouse(sel, "contextmenu", btn=2)
      b.wait_present('.contextMenu')

- The paste permission is a bit trickier. There is a [Chrome DevTools protocol API](https://chromedevtools.github.io/devtools-protocol/tot/Browser/#method-grantPermissions) for it, but it's still marked as experimental.

I was playing around with this, and https://paste.fedoraproject.org/paste/lyBZK6uS-892C33HBDTf2w is my current diff. It fixes the context menu, but it seems the permission does not make much difference. 

The patch also tries to show the error why the read/write fails, but the error message is still absolutely useless:

    error: writeText finish: Text could not be copied, use Ctrl+Insert: 
    error: readText finish: Text could not be pasted, use Shift+Insert: 

Thus  the caught `error` is something non-printable.

martinpitt commented:
> Thus the caught `error` is something non-printable.

I didn't find a way to make it useful with chromium, but it does work better in Firefox: With `error.toString()` it says

    readText call: Text could not be pasted, use Shift+Insert: TypeError: navigator.clipboard.readText is not a function terminal.js:35192:7 

And indeed the `navigator.clipboard` object only has `writeText()`, not `readText()` (you already mentioned that). For the record, Shift+Insert works fine!

Now the test actually works reasonably well (aside from the missing selection) when I run it noninteractively. For comparison, it fails when I comment out the `Browser.grantPermissions` call, so this actually does work as intended. The copy&paste does not work when I run the test interactively (`TEST_CDP_PORT` with an interactive chromium), but that's secondary.

I pushed what I have to your branch, but it still needs some work. (Thus triggering only one test)

marusak commented:
@martinpitt you pushed logging of errors, but in chrome when I don't give browser permissions it fails with:
`Uncaught (in promise) TypeError: Cannot read property 'toString' of undefined at navigator.clipboard.readText.then.catch.e`
I put there ugly condition as chrome raises with 'undefined'. Suggestions how to do it better are welcome.

Also test finally works. We need to set up click as first. I am thinking if this should be separate commit. For now it is one commit, but I am happy to pick it up into another commit. Hope it is not gonna break anything else (lets see with tests).

And I also updated to your review.

martinpitt commented:
> you pushed logging of errors, but in chrome when I don't give browser permissions it fails with:

Eek, sorry. So the error is indeed `undefined`. This was mostly an intermediate thing to understand better *why* it fails in various scenarios. Your fix with the `e ?` conditional is fine, thanks for cleaning it up!

> We need to set up click as first. I am thinking if this should be separate commit.

Yes, it should. Earlier I would have said, make it a separate PR to see if that breaks anything unrelated. But it seems it doesn't, so just making it a separate commit in this PR is fine as well (please add a Closes #). Very nice work figuring this out! I'm glad that the test now looks rather reasonable!

As an added bonus, move the `grantPermissions` stuff into a testlib `Browser` method, so that we can reuse it someplace else. (Separate commit again). But I leave that up to you, it's fine as-is too.

martinpitt commented on pkg/lib/cockpit-components-context-menu.jsx:
```diff
@@ -0,0 +1,110 @@
+        return this.state.visible &&
+            <div ref={ ref => { this.root = ref } } className="contextMenu">
+                <div className="contextMenuOption" onClick={this.props.getText}>
+                    <div className="contextMenuName"> { copy_text } </div>
```
It's the smallest thing, but please don't make this a variable. Usually we just do this inline, i. e.

    <div className="contextMenuName">{ _("Copy") }</div>

which is a bit easier to read linearly.

martinpitt commented on pkg/lib/cockpit-components-context-menu.jsx:
```diff
@@ -0,0 +1,110 @@
+            <div ref={ ref => { this.root = ref } } className="contextMenu">
+                <div className="contextMenuOption" onClick={this.props.getText}>
+                    <div className="contextMenuName"> { copy_text } </div>
+                    <div className="contextMenuShortcut">Ctrl+Insert</div>
```
This needs to be translatable, too.

martinpitt commented on pkg/lib/cockpit-components-context-menu.jsx:
```diff
@@ -0,0 +1,110 @@
+                </div>
+                <div className="contextMenuOption" onClick={this.props.setText} >
+                    <div className="contextMenuName"> { paste_text } </div>
+                    <div className="contextMenuShortcut">Shift+Insert</div>
```
Dito for these two.

marusak commented:
@martinpitt I've updated the PR again (sorry for doing such poor job the first time). I've split it into 3 commits and also added test for keyboard shortcuts.

martinpitt commented on test/common/test-functions.js:
```diff
@@ -119,11 +119,17 @@ function ph_mouse(sel, type, x, y, btn, force) {
+    if (type === "dblclick")
+        detail = 2;
+    if (["mouseover", "mouseout", "mouseenter", "mouseleave", "mousemove"].indexOf(type) > -1)
+        detail = 0;
```
Wouldn't it be simpler to default to 0 and only use 1 on "click"?

martinpitt commented on test/common/testlib.py:
```diff
@@ -556,6 +556,13 @@ def copy_js_log(self, title, label=None):
+        self.cdp.invoke("Browser.grantPermissions",
+                        origin=origin,
+                        permissions=[x for x in args])
+
```
It's the smallest thing, but would you mind moving this upwards a little bit, to similar methods? Right after `ignore_ssl_certificate_errors()` sounds right, as it's also permission related.

Also, I don't understand the `[x for x in args]` -- `args` is already a list, what does this convert?

marusak replied:
> `args` is already a list

args is tuple. and by "rather safe than sorry" I did the conversion, but now I tested it and it seems it works just fine with tuples, so I will remove it.

martinpitt replied:
Right, tuple, I forgot about that. You can write `list(args)` to be on the safe side.

martinpitt commented on pkg/lib/cockpit-components-context-menu.jsx:
```diff
@@ -0,0 +1,114 @@
+
+ContextMenu.propTypes = {
+    getText: PropTypes.func,
+    setText: PropTypes.func
```
These should have `.isRequired`.

martinpitt commented on pkg/lib/context-menu.css:
```diff
@@ -0,0 +1,22 @@
+.contextMenu {
+    position: fixed;
+    z-index: 101;
```
This could use a comment: Both why it needs to be set, and explain the magic number.

martinpitt commented on test/verify/check-terminal:
```diff
@@ -98,6 +98,69 @@ PROMPT_COMMAND='printf "\\033]0;%s@%s:%s\\007" "${USER}" "${HOSTNAME%%.*}" "${PW
         wait_line(n + 1, blank_state)
         self.assertNotIn('admin', b.text(line_sel(n + 1)))
 
+        if m.image == "rhel-7-6-distropkg":
```
Please add a comment "Fixed in PR #11148"

martinpitt commented on test/verify/check-terminal:
```diff
@@ -98,6 +98,69 @@ PROMPT_COMMAND='printf "\\033]0;%s@%s:%s\\007" "${USER}" "${HOSTNAME%%.*}" "${PW
+            b.mouse(sel, "mouseup", 10 + width, 17)
+
+
+        b.grant_permissions("http://%s:%s" % (self.machine.web_address, self.machine.web_port),
```
It feels like the origin should be computed internally by `grant_permissions()` -- I don't see us ever using any other origin.

martinpitt commented on test/verify/check-terminal:
```diff
@@ -98,6 +98,69 @@ PROMPT_COMMAND='printf "\\033]0;%s@%s:%s\\007" "${USER}" "${HOSTNAME%%.*}" "${PW
+        #   used, windowsVirtualKeyCode is used and insert key has code 45.
+        #   At the same time '-' has ascii code 45.
+        b.key_press("-", 2) # Ctrl + Insert
+        b.key_press("-", 8) # Shift + Insert
```
Urgh -- let's clean this up (some other time, not this PR) to be able to say "Insert" there :-)
